File: udemy/ds-and-a/DSAndA/StacksAndQueues/StackUsingArrayAsUnderlyingDS.cs
namespace DSAndA.StacksAndQueues
{
    public class StackUsingArrayAsUnderlyingDS
    {
        private int[] stackArray;
        private int top;

        public StackUsingArrayAsUnderlyingDS(int maxSize)
        {
            this.stackArray = new int[maxSize];

            this.top = -1;
        }

        public void Push(int element)
        {
            this.stackArray[++top] = element;
        }

        public int Pop()
        {
            return this.stackArray[top--];
        }

        public int Peek()
        {
            return this.stackArray[top];
        }
    }
}

File: udemy/ds-and-a/javascript-solutions/src/StackAndQueues/StackUsingArrayAsUnderlyingDS.js
class StackUsingArrayAsUnderlyingDS {
    constructor() {
        this.stack = [];
        this.top = -1;
        this.length = 0;
    }

    // O(1)
    push(data) {
        this.stack[++this.top] = data;
        this.length++;
    }

    // O(1)
    pop() {
        const item = this.stack[this.top--];
        this.length--;

        return item;
    }

    // O(1)
    peek() {
        return this.stack[this.top];
    }

    length() {
        return this.length;
    }

    toString() {
        return this.top >= 0 ? this.stack.slice(0, this.top + 1).join(', ') : '';
    }
}

module.exports = StackUsingArrayAsUnderlyingDS;